R/ui.R

```diff
@@ -0,0 +1,212 @@
+    ## (can't directly ask for 127.0.0.1)
+    if (is.null(getOption("plumber.apiURL")) &&
+        is.null(getOption("plumber.apiHost"))) {
+      if (is.null(req$HTTP_REFERER)) {
+        # Prevent leaking host and port if option is not set
+        api_url <- character(1)
+      }
+      else {
+        # Use HTTP_REFERER as fallback
+        api_url <- req$HTTP_REFERER
+        api_url <- sub("(\\?.*)?$", "", api_url)
+        api_url <- sub("index\\.html$", "", api_url)
+        api_url <- sub(paste0("__(", paste0(names(.globals$UIs), collapse = "|"),")__/$"), "", api_url)
+      }
+    }
+
+    utils::modifyList(list(servers = list(list(url = api_url))), spec)
+
+  }
+  # http://spec.openapis.org/oas/v3.0.3#document-structure
+  # "It is RECOMMENDED that the root OpenAPI document be named: openapi.json"
+  pr$handle("GET", "/openapi.json", openapi_fun, serializer = serializer_unboxed_json())
+
+  invisible()
+}
+
+#' Mount OpenAPI Specification to a plumber router
+#' @noRd
+unmount_openapi <- function(pr) {
+
+  pr$remove_handle("GET", "/openapi.json")
+  invisible()
+
+}
+
+#' Add UI for plumber to use
+#' @param ui A list of that plumber can use to mount
+#' a UI.
+#' @details [register_ui()] is used by other packages like `swagger`.
+#' When you load these packages, it calls [register_ui()] to provide a user
+#' interface that can interpret your plumber OpenAPI Specifications.
+#'
+#' `ui` list expects the following values
+#' \describe{
+#' \item{package}{Name of the package required for the UI.}
+#' \item{name}{Name of the UI.}
+#' \item{index}{A function that returns the HTML content of the landing page of the UI.}
+#' \item{static}{A function that returns the path to the assets (images, javascript, css, fonts) the UI will use.}
+#' }
+#' @export
+#' @rdname register_ui
+register_ui <- function(ui) {
+
+  stopifnot(is.list(ui))
+  stopifnot(is.character(ui$package) && length(ui$package) == 1L)
+
+  if (!requireNamespace(ui$package, quietly = TRUE)) {
+    stop(ui$package, " must be installed for the ", ui$name," UI to be displayed")
+  }
+
+  stopifnot(is.character(ui$name) && length(ui$name) == 1L)
+  stopifnot(is.function(ui$static))
+  stopifnot(is.function(ui$index))
+  ui_root <- paste0("/__", ui$name, "__/")
+  ui_path <- paste0(ui_root, c("index.html", ""))
+
+  mount_ui_func <- function(pr, api_url, ...) {
+
+    ui_url <- paste0(api_url, ui_root)
+
+    # Save initial extra argument values
+    args_index <- list(...)
+
+    ui_index <- function(...) {
+      # Override with arguments provided live with URI (i.e. index.html?version=2)
+      args <- utils::modifyList(args_index, list(...))
+      # Remove default arguments req and res
+      args <- args[!(names(args) %in% c("req", "res"))]
+      do.call(ui$index, args)
+    }
+    for (path in ui_path) {
+      pr$handle("GET", path, ui_index,  serializer = serializer_html())
+    }
+    pr$mount(ui_root, PlumberStatic$new(ui$static(...)))
```

Comment:
Do this via a full subrouter (not just mounting to the original)

Reply:
Also, we should emit a message/warning if there is already something mounted here